analysis/utility_analysis.py

```
@@ -146,3 +161,74 @@ def _pack_per_partition_metrics(
+    if n < 0:
+        return 0
+    index = bisect.bisect_right(BUCKET_BOUNDS, n)
+    if index > len(BUCKET_BOUNDS):
+        return -1  # todo
+    return BUCKET_BOUNDS[index]
+
+
+def _unnest_metrics(
+    metrics: List[metrics.PerPartitionMetrics]
+) -> Iterable[Tuple[Any, metrics.PerPartitionMetrics]]:
+    """Unnest metrics from different configurations."""
+    for i, metric in enumerate(metrics):
+        yield ((i, None), metric)
+    if metrics[0].metric_errors:
+        # Emits metrics for computing histogram by partition size.
+        actual_bucket_value = metrics[0].metric_errors[0].sum
+        bucket = _get_lower_bound(actual_bucket_value)
+        for i, metric in enumerate(metrics):
+            yield ((i, bucket), metric)
+
+
+def _group_utility_reports(
+        configuration_index: int,
+        reports: List[metrics.UtilityReport]) -> metrics.UtilityReport:
+    """Groups utility reports for one configuration.
+
+    'reports' contains the global report, i.e. which corresponds to all
+    partitions and reports that corresponds to partitions of some size range.
+    This function creates UtilityReportHistogram from reports by size range and
+    sets it to the global report.
+    """
+    global_report = None
+    histogram_reports = []
+    for lower_bucket_bound, report in reports:
+        #  Apache Beam does not allow input data to be changed during a map
+        #  stage. So 'report' has to be copied.
+        report = copy.deepcopy(report)
+        report.configuration_index = configuration_index
+        if lower_bucket_bound is None:
+            # only the report which corresponds to all partitions does not have
+            # the bucket.
+            global_report = report
+        else:
+            histogram_reports.append((lower_bucket_bound, report))
+    if global_report is None:
+        # it should not happen, but it better to process gracefully in case
+        # if it happens and return None.
+        return None
```

Maybe we should throw an exception then?